c/tskit/trees.c

```diff
@@ -4685,3 +4729,278 @@ tsk_tree_kc_distance(tsk_tree_t *self, tsk_tree_t *other, double lambda, double
+    memset(&t2, 0, sizeof(tsk_tree_t));
+
+    if (self->tables->sequence_length != other->tables->sequence_length) {
+        ret = TSK_ERR_SEQUENCE_LENGTH_MISMATCH;
+        goto out;
+    }
+
+    ret = check_kc_distance_samples_inputs(self, other);
+    if (ret != 0) {
+        goto out;
+    }
+
+    ret = tsk_tree_init(&t1, self, TSK_SAMPLE_LISTS);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_tree_init(&t2, other, TSK_SAMPLE_LISTS);
+    if (ret != 0) {
+        goto out;
+    }
+
+    assert(tsk_tree_first(&t1) == 1);
+    while (tsk_tree_next(&t2)) {
+        while (t1.right < t2.right) {
+            ret = check_kc_distance_tree_inputs(&t1, &t2);
+            if (ret != 0) {
+                goto out;
+            }
+            assert(tsk_tree_next(&t1) == 1);
+        }
+        ret = check_kc_distance_tree_inputs(&t1, &t2);
+        if (ret != 0) {
+            goto out;
+        }
+    }
+
+out:
+    tsk_tree_free(&t1);
+    tsk_tree_free(&t2);
+    return ret;
+}
+
+static void
+update_in_out_subtree_pairs(
+    tsk_tree_t *self, kc_vectors *kc, tsk_edge_t *e, double root_time)
+{
+    tsk_id_t edge_c, c, p, sib;
+    tsk_size_t depth;
+    double *times = self->tree_sequence->tables->nodes.time;
+
+    edge_c = e->child;
+    depth = tsk_tree_get_depth(self, e->parent);
+    for (c = e->child, p = e->parent; p != TSK_NULL; c = p, p = self->parent[p]) {
+        for (sib = self->left_child[p]; sib != TSK_NULL; sib = self->right_sib[sib]) {
+            if (sib != c) {
+                update_kc_vectors_all_pairs(
+                    self, kc, edge_c, sib, depth, root_time - times[p]);
+            }
+        }
+        depth--;
+    }
+}
+
+static int
+update_in_subtree_pairs(tsk_tree_t *t, kc_vectors *kc, tsk_id_t u, double root_time)
+{
+    int stack_top;
+    tsk_size_t depth;
+    struct kc_stack_elmt *stack;
+    tsk_id_t v, c1, c2;
+    double *times;
+    int ret = 0;
+
+    stack = malloc(t->num_nodes * sizeof(*stack));
+    if (stack == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    times = t->tree_sequence->tables->nodes.time;
+
+    stack_top = 0;
+    stack[stack_top].node = u;
+    stack[stack_top].depth = tsk_tree_get_depth(t, u);
+    while (stack_top >= 0) {
+        v = stack[stack_top].node;
+        depth = (tsk_size_t) stack[stack_top].depth;
+        stack_top--;
+
+        if (!tsk_tree_is_sample(t, v)) {
+            for (c1 = t->left_child[v]; c1 != TSK_NULL; c1 = t->right_sib[c1]) {
+                stack_top++;
+                stack[stack_top].node = c1;
+                stack[stack_top].depth = depth + 1;
+
+                for (c2 = t->right_sib[c1]; c2 != TSK_NULL; c2 = t->right_sib[c2]) {
+                    update_kc_vectors_all_pairs(
+                        t, kc, c1, c2, depth, root_time - times[v]);
+                }
+            }
+        }
+    }
+
+out:
+    tsk_safe_free(stack);
+    return ret;
+}
+
+static int
```

Comment:
We should probably group these ``tsk_tree_x`` methods with the rest of tsk_tree_x functions within the file for consistency.